runtime/vcache/loader.go

```diff
@@ -0,0 +1,589 @@
+package vcache
+
+import (
+	"fmt"
+	"io"
+	"net/netip"
+	"sync"
+
+	"github.com/brimdata/zed"
+	"github.com/brimdata/zed/vector"
+	"github.com/brimdata/zed/vng"
+	"github.com/brimdata/zed/zcode"
+	"github.com/brimdata/zed/zson"
+	"golang.org/x/sync/errgroup"
+)
+
+// loader handles loading vector data on demand for only the fields needed
+// as specified in the projection.  Each load is executed with a multiphase
+// process: first, we build a mirror of the vng metadata where each node has a
+// lock and places to store the bulk data so that it may be reused across
+// projections.  This is called the shadow object.  Then, we fill in the shadow
+// with data vectors dynamically and create runtime vectors as follows:
+//
+//	(1) create a mirror data structure (shadow)
+//	(2) concurrently load all the nulls and tags/lens/etc that will be needed (fetchNulls)
+//	(3) compute top-down flattening of nulls (concurrently) (flatten)
+//	(4) load all data that is projected using the nulls to flatten any unloaded data (fetchVals)
+//	(5) form a projection from the fully loaded data nodes (project)
+//
+// The zctx passed into the loader is dynamic and comes from each query context that
+// uses the vcache.  No zctx types are stored in the shadow (except for primitive types
+// in shadowed vector.Any primitives that are shared).  We otherwise allocate all
+// vector.Any zed.Types using the passed-in zctx.
+type loader struct {
+	zctx *zed.Context
+	r    io.ReaderAt
+}
+
+// Load all vector data into the in-memory shadow that is needed and not yet loaded
+// and return a new vector.Any using the data vectors in cache.  This may be called
+// concurrently on the same shadow and fine-grained locking insures that any given
+// data vector is loaded just once and such loads may be executed concurrently (even
+// when only one thread is calling load).  If paths is nil, then the entire value
+// is loaded.
+func (l *loader) load(paths Path, s shadow) (vector.Any, error) {
```

Comment:
Nit: Shadow feels like it should come first because it's the thing from which you're loading. The paths just says what parts of the shadow to load.  (Same goes for all the other functions and methods that take a shadow and paths.)
```suggestion
func (l *loader) load(s shadow, paths Path) (vector.Any, error) {
```

Reply:
I'm gonna leave this one